tests/unit/connection.js

```diff
@@ -0,0 +1,228 @@
+const config = require('../data/config')
+const Connection = require('../../lib/connection')
+const helper = require('../helper')
+const sinon = require('sinon')
+
+exports['static - withConnection'] = {
+  setUp (done) {
+    helper.initDatabase.call(this).then(() => done())
+  },
+
+  tearDown (done) {
+    this.connection.end()
+    done()
+  },
+
+  'error with connect' (test) {
+    const providedError = new Error()
+    sinon.stub(Connection.prototype, 'connect').returns(Promise.reject(providedError))
+
+    Connection.withConnection(config, (connection) => {
+      test.ok(false, 'should not invoke callback')
+    })
+      .catch((error) => {
+        test.strictEqual(error, providedError, 'should propagate error')
+
+        Connection.prototype.connect.restore()
+        test.done()
+      })
+  },
+
+  'error with callback' (test) {
+    const providedError = new Error()
+    sinon.spy(Connection.prototype, 'end')
+
+    Connection.withConnection(config, (connection) => {
+      test.ok(connection instanceof Connection, 'should provide connection')
+      test.strictEqual(Connection.prototype.end.callCount, 0,
+        'end() should not be called before callback')
+      return Promise.reject(providedError)
+    })
+      .catch((error) => {
+        test.strictEqual(error, providedError, 'should propagate error')
+        test.strictEqual(Connection.prototype.end.callCount, 1,
+          'end() should be called after callback')
+
+        Connection.prototype.end.restore()
+        test.done()
+      })
+  },
+
+  'error with end, resolved callback' (test) {
+    const providedValue = {}
+    const end = Connection.prototype.end
+    sinon.stub(Connection.prototype, 'end', function () {
```

Comment:
Could this be an arrow function, or do you need function scope here for your `.call` below?